src/platform/packages/shared/kbn-coloring/src/shared_components/color_mapping/color/color_assignment_matcher.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { RawValue, deserializeField } from '@kbn/data-plugin/common';
+import { ColorMapping } from '../config';
+
+type AssignmentMatchCount = [assignmentIndex: number, matchCount: number];
+
+/**
+ * A class to encapsulate assignment logic
+ */
+export class ColorAssignmentMatcher {
+  /**
+   * Reference to original assignments
+   */
+  readonly #assignments: ColorMapping.Assignment[];
+
+  /**
+   * Map values (or keys) to assignment index and match count
+   */
+  #assignmentMap: Map<string, AssignmentMatchCount>;
+
+  constructor(assignments: ColorMapping.Assignment[]) {
+    this.#assignments = assignments;
+    this.#assignmentMap = this.#assignments.reduce<Map<string, AssignmentMatchCount>>(
+      (acc, assignment, i) => {
+        assignment.rules.forEach((rule) => {
```

Comment:
We can also consider flattening the Assignment array into something more quick to run: Array<{rule:Rule, color: c}> where you have a single rule associated to a single color, and not multiple rules associated to a single color. in the end we just pick up the color from the first or the last matching rule

Reply:
Are you talking about changing the `ColorMapping.Assignment` type or the `#assignments` locally to this class/file to simplify the `reduce`?

Reply:
don't worry, nothing to consider for now